File: DSA/PriyankaSingh_590016827/day40question3.c

#include <stdio.h>

#define MAX 100

int stack[MAX], maxStack[MAX];
int top = -1, maxTop = -1;

void push(int x) {
    stack[++top] = x;
    if (maxTop == -1 || x >= maxStack[maxTop])
        maxStack[++maxTop] = x;
}

void pop() {
    if (top == -1) return;
    if (stack[top] == maxStack[maxTop])
        maxTop--;
    top--;
}

int getMax() {
    if (maxTop == -1) return -1;
    return maxStack[maxTop];
}

int main() {
    push(3);
    push(5);
    push(2);
    printf("Max: %d\n", getMax());

    pop();
    printf("Max after pop: %d\n", getMax());

    return 0;
}

File: DSA/VenikaPanwar_590013755/Day_40_Question_3.java
import java.util.Scanner;
import java.util.Stack;

class MaxStack {
    Stack<Integer> main = new Stack<>();
    Stack<Integer> maxStack = new Stack<>();

    public void push(int x) {
        main.push(x);
        if (maxStack.isEmpty() || x >= maxStack.peek())
            maxStack.push(x);
    }

    public void pop() {
        if (!main.isEmpty()) {
            int top = main.pop();
            if (top == maxStack.peek()) {
                maxStack.pop();
            }
        }
    }

    public int top() {
        return main.isEmpty() ? -1 : main.peek();
    }

    public int getMax() {
        return maxStack.isEmpty() ? -1 : maxStack.peek();
    }
}

public class MaxStackTest {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        MaxStack ms = new MaxStack();

        ms.push(3);
        ms.push(5);
        ms.push(2);
        System.out.println("Current Max: " + ms.getMax()); // 5

        ms.pop();
        System.out.println("Current Max: " + ms.getMax()); // 5

        ms.pop();
        System.out.println("Current Max: " + ms.getMax()); // 3
    }
}
